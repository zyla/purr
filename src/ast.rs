use crate::symbol::Symbol;

#[derive(Debug)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Located<T>(pub SourceSpan, pub T);

#[derive(Debug)]
pub struct Comment(pub String);

#[derive(Debug)]
pub struct Commented<T>(pub Vec<Comment>, pub T);

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Debug)]
pub struct ModuleInner {
    pub name: Symbol,
    pub exports: Option<Vec<DeclarationRef>>,
    pub declarations: Vec<Declaration>,
}

#[derive(Debug)]
pub enum DeclarationRef {
    // TODO
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Debug)]
pub enum DeclarationKind {
    ValueDeclaration(ValueDeclaration),
}

#[derive(Debug)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind

    // TODO
    // pub binders: Vec<Binder>
    pub expr: Vec<GuardedExpr>,
}

#[derive(Debug)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Debug)]
pub enum Guard {
    // TODO
}

pub type Expr = Located<ExprKind>;

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal<Expr>),
    // TODO
}

#[derive(Debug)]
pub enum Literal<T> {
    Integer(i32),
    Float(f64),
    String(String),
    Char(char),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}
