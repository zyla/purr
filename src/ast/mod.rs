pub mod declarations;
pub mod expr;
pub mod meta;
pub mod types;

use crate::ModuleId;
pub use declarations::*;
pub use expr::*;
pub use meta::*;
pub use types::*;

use crate::symbol::Symbol;

#[salsa::interned]
pub struct AbsoluteName {
    pub module: ModuleId,
    pub name: Symbol,
}

#[salsa::interned]
pub struct QualifiedName {
    pub module: Option<ModuleId>,
    pub name: Symbol,
}

impl QualifiedName {
    pub fn is_actually_qualified(&self, db: &dyn crate::Db) -> bool {
        self.module(db).is_some()
    }

    pub fn new_unqualified(db: &dyn crate::Db, name: Symbol) -> Self {
        Self::new(db, None, name)
    }

    pub fn new_qualified(db: &dyn crate::Db, module: ModuleId, name: Symbol) -> Self {
        Self::new(db, Some(module), name)
    }
}
