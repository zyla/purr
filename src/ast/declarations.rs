use super::{Commented, Located, Pat, PossiblyGuardedExpr, Type};
use crate::ast::QualifiedName;
use crate::symbol::Symbol;

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ModuleInner {
    pub name: ModuleName,
    pub exports: Option<Vec<DeclarationRef>>,
    pub imports: Vec<Import>,
    pub declarations: Vec<Declaration>,
}

pub type Import = Located<ImportInner>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ImportInner {
    pub module: ModuleName,
    pub kind: ImportDeclarationKind,
    pub alias: Option<ModuleName>,
}

pub type DeclarationRef = Located<DeclarationRefKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum NameSource {
    UserNamed,
    CompilerNamed,
}

pub type ModuleName = QualifiedName;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefKind {
    TypeClass {
        name: Symbol,
    },
    TypeOp {
        name: Symbol,
    },
    Type {
        name: Symbol,
        constructors: Option<DeclarationRefConstructors>,
    },
    Value {
        name: Symbol,
    },
    ValueOp {
        name: Symbol,
    },
    TypeInstanceRef {
        name: Symbol,
        name_source: NameSource,
    },
    Module {
        name: ModuleName,
    },
    ReExport {
        imported_from: ModuleName,
        defined_in: ModuleName,
        declaration_ref: Box<DeclarationRef>,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefConstructors {
    All,
    Some(Vec<Symbol>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ImportDeclarationKind {
    Implicit,
    Explicit(Vec<DeclarationRef>),
    Hiding(Vec<DeclarationRef>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct RoleDeclarationData {
    pub ident: Symbol,
    pub role: Vec<Role>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Role {
    Nominal,
    Representational,
    Phantom,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeDeclarationData {
    pub ident: Symbol,
    pub r#type: Type,
}

impl TypeDeclarationData {
    pub fn new(ident: Symbol, r#type: Type) -> Self {
        Self { ident, r#type }
    }
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationKind {
    Data {
        type_: DataDeclType,
        name: Symbol,
        params: Vec<TypeParameter>,
        kind: Option<Type>,
        constructors: Vec<DataConstructorDeclaration>,
    },

    // TODO: do we need this? seems internal
    DataBindingGroup(Vec<Declaration>),

    TypeSynonym {
        name: Symbol,
        params: Vec<TypeParameter>,
        body: Type,
    },

    KindSignature {
        for_type: KindSignatureFor,
        name: Symbol,
        kind: Kind,
    },

    Role(RoleDeclarationData),

    TypeSignature(TypeDeclarationData),

    ValueDeclaration(ValueDeclaration),

    // C x y = z, used in `let`
    Destructuring {
        pat: Pat,
        expr: PossiblyGuardedExpr,
    },

    ForeignValue {
        name: Symbol,
        type_: Type,
    },

    Class(TypeClassDeclaration),

    InstanceChain(Vec<InstanceDeclaration>),

    Operator {
        associativity: Associativity,
        precedence: u8,
        name: OperatorTarget,
        operator: Symbol,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Associativity {
    None,
    Left,
    Right,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum OperatorTarget {
    Type(QualifiedName),
    DataConstructor(QualifiedName),
    Value(QualifiedName),
}

pub type Kind = Type;

pub type TypeParameter = (Symbol, Option<Kind>);

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeClassDeclaration {
    pub constraints: Vec<Type>,
    pub name: Symbol,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct InstanceDeclaration {
    pub constraints: Vec<Type>,
    pub instance_type: InstanceType,
    pub instance_name: Option<Symbol>,
    pub class: QualifiedName,
    pub args: Vec<Type>,
    pub body: Vec<Declaration>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct Fundep {
    pub from: Vec<Symbol>,
    pub to: Vec<Symbol>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InstanceType {
    Plain,
    Derive,
    DeriveNewtype,
}

// Note: `data` and `newtype` signatures are actually declarations without constructors
#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum KindSignatureFor {
    TypeSynonym,
    Class,
}

pub type DataConstructorDeclaration = Located<Commented<DataConstructorDeclarationData>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct DataConstructorDeclarationData {
    pub name: Symbol,

    // TODO: in original AST they have names. Why? I thought datacon fields are unnamed in PS
    pub fields: Vec<Type>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DataDeclType {
    Data,
    ForeignData,
    Newtype,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind
    pub params: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}
