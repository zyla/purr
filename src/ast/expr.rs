use super::{Declaration, Located, Type};
use crate::ast::QualifiedName;
use crate::string::PSChar;
use crate::string::PSString;
use crate::symbol::Symbol;
use ordered_float::OrderedFloat;

pub type Expr = Located<ExprKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ExprKind {
    Literal(Literal<Expr>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Expr>, Vec<(InfixOp, Expr)>),

    /// Record field accessor
    Accessor(Box<Expr>, Symbol),

    RecordUpdate(Box<Expr>, RecordUpdate),

    Var(QualifiedName),

    /// Standalone operator
    Operator(InfixOp),

    DataConstructor(QualifiedName),

    App(Box<Expr>, Vec<Expr>),

    Lam(Vec<Pat>, Box<Expr>),

    Case {
        exprs: Vec<Expr>,
        branches: Vec<CaseBranch>,
    },

    If {
        cond: Box<Expr>,
        then_: Box<Expr>,
        else_: Box<Expr>,
    },

    Typed(Box<Expr>, Box<Type>),

    Let {
        decls: Vec<Declaration>,
        body: Box<Expr>,
    },

    Wildcard,

    // Pseudo-expression, used only as an intermediate value during parsing.
    RecordUpdateSuffix(RecordUpdate),

    // Pseudo-expression, used only as an intermediate value during parsing.
    NamedPat(Symbol, Box<Expr>),

    Do(Vec<DoItem>),

    Ado(Vec<DoItem>, Box<Expr>),

    Negate(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InfixOp {
    Symbol(QualifiedName),
    Backtick(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DoItem {
    Let(Vec<Declaration>),
    Expr(Expr),
    Bind(Pat, Expr),
}

type RecordUpdate = Vec<(Symbol, Expr)>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum RecordLiteralOrUpdate {
    Literal(Vec<(Symbol, Expr)>),
    Update(Vec<(Symbol, Expr)>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct CaseBranch {
    pub pats: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PossiblyGuardedExpr {
    Unconditional(Expr),
    Guarded(Vec<GuardedExpr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Guard {
    Expr(Expr),
    Bind(Pat, Expr),
}

pub type Pat = Located<PatKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PatKind {
    Literal(Literal<Pat>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Pat>, Vec<(QualifiedName, Pat)>),

    Var(Symbol),

    DataConstructorApp(QualifiedName, Vec<Pat>),

    Wildcard,

    Named(Symbol, Box<Pat>),

    Typed(Box<Pat>, Box<Type>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Literal<T> {
    Integer(i64),
    Float(OrderedFloat<f64>),
    String(PSString),
    Char(PSChar),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}

#[test]
fn test_size() {
    // Note: it was 56 before adding multi-case.
    assert_eq!(std::mem::size_of::<Expr>(), 72);
}
