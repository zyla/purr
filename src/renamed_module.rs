use crate::ast::AbsoluteName;
use crate::ast::DeclarationRefConstructors;
use crate::indexed_module::TypeClassDecl;
use crate::indexed_module::TypeDecl;
use fxhash::FxHashMap;
use std::collections::{HashMap, HashSet};

use salsa::DebugWithDb;

use petgraph::{algo::tarjan_scc, prelude::DiGraph};

use crate::{
    ast::{Declaration, DeclarationRefKind, ImportDeclarationKind},
    indexed_module::{IndexedModule, ValueDecl},
    rename::rename_module,
    symbol::Symbol,
    Db, ModuleId, ParsedModule,
};

#[salsa::interned]
pub struct DeclId {
    pub namespace: Namespace,
    pub module: ModuleId,
    pub name: Symbol,
}

impl DeclId {
    fn to_absolute_name(self, db: &dyn Db) -> AbsoluteName {
        AbsoluteName::new(db, self.module(db), self.name(db))
    }
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb, Hash)]
pub enum Namespace {
    Class,
    Type,
    Value,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct RenamedModule {
    pub module_id: ModuleId,
    pub imported: Vec<(Option<ModuleId>, DeclId)>,
    pub exported: Vec<DeclId>,
    pub types: FxHashMap<AbsoluteName, TypeDecl>,
    pub values: FxHashMap<AbsoluteName, ValueDecl>,
    pub classes: FxHashMap<AbsoluteName, TypeClassDecl>,
}

#[salsa::tracked]
pub fn renamed_value_decl(db: &dyn Db, id: DeclId) -> ValueDecl {
    assert!(id.namespace(db) == Namespace::Value);
    // FIXME: we shoudn't have to clone here
    let abs_name = id.to_absolute_name(db);
    renamed_module(db, id.module(db))
        .values
        .get(&abs_name)
        .cloned()
        .unwrap_or_else(|| {
            panic!(
                "renamed_value_decl for nonexistent value {:?}",
                abs_name.into_debug_all(db)
            )
        })
}

#[salsa::tracked]
pub fn renamed_module(db: &dyn Db, module_id: ModuleId) -> RenamedModule {
    let mut indexed = crate::indexed_module::indexed_module(db, module_id);
    let mut imported = crate::renamed_module::imported_decls(db, module_id);
    let mut exported = crate::renamed_module::exported_decls(db, module_id);
    let mut diagnositics = vec![];

    let module = crate::parsed_module(db, module_id);

    rename_module(
        db,
        &mut indexed,
        &mut imported,
        &mut exported,
        &mut diagnositics,
    );

    let mut graph = DiGraph::<Declaration, ()>::new();
    let mut node_indices = HashMap::new();

    for declaration in module.ast.declarations.iter() {
        let node_index = graph.add_node(declaration.clone());
        node_indices.insert(declaration.clone(), node_index);
    }

    // TODO: Add edges between nodes based on data structure
    // for declaration in module.ast.declarations.iter() {
    //     use crate::ast::DeclarationKind;
    //     match &***declaration {
    //         // Handle each variant of DeclarationKind
    //         DeclarationKind::Data { constructors, .. } => {
    //             for constructor in constructors {
    //                 if let Some(target_node) = node_indices
    //                     .get(&DeclarationKind::ValueDeclaration(constructor.name.clone()))
    //                 {
    //                     graph.add_edge(node_indices[declaration], *target_node, ());
    //                 }
    //             }
    //         }
    //         DeclarationKind::TypeSynonym { body, .. } => todo!(),
    //     }
    // }

    let _scc = tarjan_scc(&graph);

    RenamedModule {
        module_id,
        imported: imported.clone(),
        exported: exported.clone(),
        values: indexed.values,
        types: indexed.types,
        classes: indexed.classes,
    }
}

struct ExportedDeclExtractor<'a> {
    db: &'a dyn Db,
    module_id: ModuleId,
    exported_decls: Vec<DeclId>,
    imported_decls: Vec<(Option<ModuleId>, DeclId)>,
}

impl<'a> ExportedDeclExtractor<'a> {
    fn extract(&mut self, module: &ParsedModule, indexed: &IndexedModule) {
        let db = self.db;
        match &module.ast.exports {
            Some(decl_ref_kind) => {
                for ref_decl in decl_ref_kind {
                    use DeclarationRefKind::*;

                    match &**ref_decl {
                        Module { name: qualified_as } => {
                            self.exported_decls.extend(
                                self.imported_decls
                                    .iter()
                                    .filter(|(name, _)| {
                                        name.is_some_and(|name| name == *qualified_as)
                                    })
                                    .map(|(_, decl)| *decl),
                            );
                        }
                        x => {
                            self.exported_decls
                                .extend(to_decls_id(db, self.module_id, x));
                        }
                    }
                }
            }
            None => {
                for abs_name in indexed.values.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Value,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }

                for abs_name in indexed.types.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Type,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }

                for abs_name in indexed.classes.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Class,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }
            }
        }
    }
}

#[salsa::tracked]
pub fn exported_decls(db: &dyn Db, module_id: ModuleId) -> Vec<DeclId> {
    let indexed = crate::indexed_module::indexed_module(db, module_id);
    let module = crate::parsed_module(db, module_id);
    let imported = imported_decls(db, module_id);

    let mut extractor = ExportedDeclExtractor {
        db,
        module_id,
        exported_decls: Default::default(),
        imported_decls: imported,
    };

    extractor.extract(&module, &indexed);

    extractor.exported_decls
}

pub fn imported_decls(db: &dyn Db, module_id: ModuleId) -> Vec<(Option<ModuleId>, DeclId)> {
    let module = crate::parsed_module(db, module_id);

    let mut imports: Vec<(Option<ModuleId>, DeclId)> = Vec::new();

    let prim_module_id = ModuleId::new(db, "Prim".into());

    let has_prim = module
        .ast
        .imports
        .iter()
        .any(|i| i.module == prim_module_id);

    if module_id != prim_module_id && !has_prim {
        crate::renamed_module::exported_decls(db, prim_module_id)
            .iter()
            .for_each(|i| imports.push((None, *i)));
    }

    for import in &module.ast.imports {
        use ImportDeclarationKind::*;
        match &import.kind {
            Implicit => {
                crate::renamed_module::exported_decls(db, import.module)
                    .iter()
                    .for_each(|i| imports.push((import.alias, *i)));
            }
            Explicit(decls) => {
                decls
                    .iter()
                    .flat_map(|i| to_decls_id(db, import.module, i))
                    .for_each(|i| imports.push((import.alias, i)));
            }
            Hiding(decls) => {
                let excluded: HashSet<DeclId> = decls
                    .iter()
                    .map(|i| to_decl_id(db, import.module, i))
                    .collect();
                crate::renamed_module::exported_decls(db, import.module)
                    .iter()
                    .filter(|i| !excluded.contains(i))
                    .for_each(|i| imports.push((import.alias, *i)));
            }
        }
    }

    imports
}

fn to_decls_id(db: &dyn Db, module_id: ModuleId, kind: &DeclarationRefKind) -> Vec<DeclId> {
    use DeclarationRefKind::*;

    match kind {
        Type { name, constructors } => {
            let mut decls = vec![DeclId::new(db, Namespace::Type, module_id, *name)];

            if let Some(constructors) = constructors {
                match constructors {
                    DeclarationRefConstructors::Some(constructors) => {
                        for constructor_name in constructors {
                            decls.push(DeclId::new(
                                db,
                                Namespace::Value,
                                module_id,
                                *constructor_name,
                            ));
                        }
                    }
                    DeclarationRefConstructors::All => {}
                }
            }
            decls
        }
        kind => vec![to_decl_id(db, module_id, kind)],
    }
}

fn to_decl_id(db: &dyn Db, module_id: ModuleId, kind: &DeclarationRefKind) -> DeclId {
    use DeclarationRefKind::*;

    match *kind {
        TypeClass { name } => DeclId::new(db, Namespace::Class, module_id, name),
        TypeOp { name } => DeclId::new(db, Namespace::Type, module_id, name),
        Type { name, .. } => DeclId::new(db, Namespace::Type, module_id, name),
        Value { name } => DeclId::new(db, Namespace::Value, module_id, name),
        ValueOp { name } => DeclId::new(db, Namespace::Value, module_id, name),
        TypeInstanceRef { name, .. } => DeclId::new(db, Namespace::Type, module_id, name),
        Module { .. } => panic!("Cannot map module to DeclId"),
    }
}

#[cfg(test)]
mod tests {
    use crate::Diagnostics;

    use super::*;
    use crate::utils::tests::DropSalsaId;
    use indoc::indoc;
    use insta::{self, assert_snapshot};

    const LIB1: &str = indoc!(
        "
        module Lib where

        data Foo = Bar
        "
    );

    const LIB2: &str = indoc!(
        "
        module Lib2 where
         
        x = 1
        y = 2
        "
    );

    fn export_decls(input: &str, deps: Vec<&str>) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = ModuleId::new(db, "Test".into());

        deps.into_iter().zip(1..).for_each(|(deb, i)| {
            db.add_source_file(format!("Lib{}.purs", i).into(), deb.into())
                .unwrap();
        });

        format!(
            "{:#?}",
            (
                exported_decls(db, module_id).into_debug_all(db),
                renamed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    fn import_decls(input: &str, deps: Vec<&str>) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = ModuleId::new(db, "Test".into());
        deps.into_iter().zip(1..).for_each(|(deb, i)| {
            db.add_source_file(format!("Lib{}.purs", i).into(), deb.into())
                .unwrap();
        });

        format!(
            "{:#?}",
            (
                imported_decls(db, module_id).into_debug_all(db),
                renamed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    #[test]
    fn export_data_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (Foo) where
    data Foo
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn export_blanket_data_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test where
    data Foo
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn reexport_subset() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib2 (x) as X
    "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_all() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib2 as X
    "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_export_list() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib as X
    "
            ),
            vec![indoc!(
                "
            module Lib (x) where
            x = 1
            y = 2
            "
            )]
        ))
    }

    #[test]
    #[ignore = "Type classes are not yet supported"]
    fn export_class_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test where

    class Foo a where
      foo :: a
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn import_data_decl() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Foo (Foo(..))
        "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn import_data_qualified_decl() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib as Lib
        "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn import_all_fns() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib2
        "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn import_subset() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib2 (x)
        "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_qualified() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test
              ( module L
              , x ) where
            
            import Lib as L
            
            x = 1
            "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn reexport_qualified_itself() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test (module L, module Test) where
            
            import Lib as L
            
            x = 1
            "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    #[ignore = "how should we handle this?"]
    fn import_self() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test where
            import Test
            x = 1
            "
            ),
            vec![]
        ))
    }
}
