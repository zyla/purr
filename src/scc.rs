use salsa::DebugWithDb;

use crate::renamed_module::Namespace;
use crate::symbol::Symbol;

use crate::{renamed_module::DeclId, Db, ModuleId};

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb, Hash)]
pub struct SccId {
    pub namespace: Namespace,
    pub module: ModuleId,
    pub name: Symbol,
}


#[salsa::tracked]
pub fn scc_of(db: &dyn Db, decl: DeclId) -> Vec<SccId> {
   let scc = SccId {
        namespace: decl.namespace,
        module: decl.module,
        name: decl.name
    };
    vec![scc];
}
