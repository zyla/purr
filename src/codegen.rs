use crate::ast::PatKind;
use crate::ast::{Expr, ExprKind};
use fxhash::FxHashSet;

use crate::ast::AbsoluteName;
use crate::renamed_module::DeclId;
use crate::renamed_module::Namespace;
use crate::scc::scc_of;
use crate::scc::SccId;
use crate::typecheck::typechecked_scc;
use std::fmt::Write;

/// Different ways to reference `AbsoluteName`s.
#[derive(PartialEq, Eq, Hash, Debug, Clone, Copy)]
pub enum ReferenceMode {
    /// Names are imported from their respective modules. `Var(AbsoluteName)` is compiled to
    /// `Some_Module.valueName`.
    ModuleImport,
    /// Names are imported from their respective modules. `Var(AbsoluteName)` is compiled to
    /// `Some_Module__valueName`.
    LocalConstant,
}

#[salsa::accumulator]
pub struct CodeAccumulator(pub String);

/// Generate code for the given value and its transitive dependencies, collecting it in the
/// `CodeAccumulator` accumulator.
#[salsa::tracked]
pub fn value_decl_code_acc(db: &dyn crate::Db, id: AbsoluteName) {
    scc_code_acc(
        db,
        scc_of(
            db,
            DeclId::new(db, Namespace::Value, id.module(db), id.name(db)),
        ),
    )
}

macro_rules! cg_write {
    ($cg:expr, $($arg:tt)*) => {
        write!(&mut $cg.code_buffer, $($arg)*).expect("formatting to a String shouldn't fail")
    };
}

/// Generate code for the given SCC and its transitive dependencies, collecting it in the
/// `CodeAccumulator` accumulator.
#[salsa::tracked]
pub fn scc_code_acc(db: &dyn crate::Db, id: SccId) {
    let mut cg = CodeGenerator {
        db,
        reference_mode: ReferenceMode::LocalConstant,
        code_buffer: String::new(),
        dependencies: Default::default(),
    };
    for (id, expr, _) in typechecked_scc(db, id) {
        cg_write!(
            cg,
            "const {}__{} = ",
            id.module(db).name(db),
            id.name(db).text(db)
        );
        cg.expr(&expr);
        cg_write!(cg, ";\n");
    }
    for dep in cg.dependencies {
        value_decl_code_acc(db, dep);
    }
    CodeAccumulator::push(db, cg.code_buffer);
}

struct CodeGenerator<'d> {
    db: &'d dyn crate::Db,
    reference_mode: ReferenceMode,
    code_buffer: String,
    dependencies: FxHashSet<AbsoluteName>,
}

impl<'d> CodeGenerator<'d> {
    fn expr(&mut self, e: &Expr) {
        let db = self.db;
        use ExprKind::*;

        match &**e {
            Var(v) => match v.to_absolute_name(db) {
                Some(abs) => {
                    self.dependencies.insert(abs);
                    match self.reference_mode {
                        ReferenceMode::ModuleImport => {
                            cg_write!(
                                self,
                                "{}.{}",
                                abs.module(db).name(db),
                                abs.name(db).text(db)
                            );
                        }
                        ReferenceMode::LocalConstant => {
                            cg_write!(
                                self,
                                "{}__{}",
                                abs.module(db).name(db),
                                abs.name(db).text(db)
                            );
                        }
                    }
                }
                None => {
                    cg_write!(self, "{}", v.name(db).text(db));
                }
            },
            App(f, xs) => {
                cg_write!(self, "(");
                self.expr(f);
                cg_write!(self, ")(");
                for (i, x) in xs.iter().enumerate() {
                    if i > 0 {
                        cg_write!(self, ", ");
                    }
                    self.expr(x);
                }
                cg_write!(self, ")");
            }
            Lam(pats, body) => {
                cg_write!(self, "(");
                for (i, p) in pats.iter().enumerate() {
                    if i > 0 {
                        cg_write!(self, ", ");
                    }
                    match &**p {
                        PatKind::Var(v) => {
                            cg_write!(self, "{}", v.text(db));
                        }
                        _ => panic!("invalid desugared lambda pattern {:?}", p),
                    }
                }
                cg_write!(self, ") => ");
                self.expr(body);
            }
            Literal(crate::ast::Literal::Integer(x)) => {
                cg_write!(self, "{}", x);
            }
            Error => {
                cg_write!(
                    self,
                    "((() => {{ throw new Error('Reached broken code') }})())",
                );
            }
            _ => todo!("codegen: unsupported expr {:?}", e),
        }
    }
}

/// Generate code for the given SCC without its dependencies. For use in multi-module code
/// generation mode.
#[salsa::tracked]
pub fn scc_code(db: &dyn crate::Db, id: SccId) -> String {
    let mut cg = CodeGenerator {
        db,
        reference_mode: ReferenceMode::ModuleImport,
        code_buffer: String::new(),
        dependencies: Default::default(),
    };
    for (id, expr, _) in typechecked_scc(db, id) {
        cg_write!(cg, "const {} = ", id.name(db).text(db));
        cg.expr(&expr);
        cg_write!(cg, ";\n");
    }
    cg.code_buffer
}

#[cfg(test)]
mod bundle_tests {
    use super::*;
    use crate::symbol::Symbol;
    use crate::ModuleId;
    use indoc::indoc;
    use insta::*;

    fn test_bundle(inputs: &[&str], entrypoint: (&str, &str)) -> String {
        let _ = env_logger::builder().is_test(true).try_init();
        let db = &mut crate::Database::new();

        inputs.into_iter().zip(1..).for_each(|(source, i)| {
            db.add_source_file(format!("Lib{}.purs", i).into(), source.to_string())
                .unwrap();
        });

        value_decl_code_acc::accumulated::<CodeAccumulator>(
            db,
            AbsoluteName::new(
                db,
                ModuleId::new(db, entrypoint.0.to_string()),
                Symbol::new(db, entrypoint.1.to_string()),
            ),
        )
        .join("")
    }

    #[test]
    fn simple1() {
        assert_snapshot!(test_bundle(
            &[indoc!(
                r"
                module Test where
                -- TODO: autoimport Prim
                import Prim (Int)
                foo :: Int
                foo = 1
                "
            )],
            ("Test", "foo")
        ));
    }

    #[test]
    fn simple_dep() {
        assert_snapshot!(test_bundle(
            &[indoc!(
                r"
                module Test where
                -- TODO: autoimport Prim
                import Prim (Int)
                unused :: Int
                unused = 1
                foo :: Int
                foo = answer
                answer :: Int
                answer = 42
                "
            )],
            ("Test", "foo")
        ));
    }

    #[test]
    #[ignore = "Can't typecheck lambda with type signature yet"]
    fn function_call() {
        assert_snapshot!(test_bundle(
            &[indoc!(
                r"
                module Test where
                -- TODO: autoimport Prim
                import Prim (Int)
                foo :: Int
                foo = frob answer
                answer :: Int
                answer = 42
                frob :: Int -> Int
                -- note: explicit lambda because no equations desugaring yet
                frob = \x -> x
                "
            )],
            ("Test", "foo")
        ));
    }
}
