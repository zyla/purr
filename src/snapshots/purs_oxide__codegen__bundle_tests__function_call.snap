---
source: src/codegen.rs
expression: "test_bundle(&[indoc!(r\"\n                module Test where\n                -- TODO: autoimport Prim\n                import Prim (Int)\n                foo :: Int\n                foo = frob answer\n                answer :: Int\n                answer = 42\n                frob :: Int -> Int\n                -- note: explicit lambda because no equations desugaring yet\n                frob = \\x -> x\n                \")],\n    (\"Test\", \"foo\"))"
---
const Test__foo = (Test__frob)(Test__answer);
const Test__frob = (x) => x;
const Test__answer = 42;
